### PR TITLE
[State Migration] Support stop migration (writing old/new DB)

### PR DESCRIPTION
**Pull request by KimKyungup**
## Proposed changes
This PR introduce the state trie batch write for old/new DB.
This makes users to be able to stop state migration.
Also during migration, if there is an error case, it can recover the old database.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: blockchain/state_migration.go [hunk 1]
```diff
@@ -84,10 +84,14 @@ func (bc *BlockChain) concurrentRead(db *statedb.Database, quitCh chan struct{},
...
 }
 
-func (bc *BlockChain) migrateState(rootHash common.Hash) error {
+func (bc *BlockChain) migrateState(rootHash common.Hash) (returnErr error) {
 	bc.wg.Add(1)
 	defer bc.wg.Done()
 
+	defer func() {
+		bc.db.FinishStateMigration(returnErr == nil)
+	}()
```
Comment (aidan-kwon) [inline blockchain/state_migration.go:93]:
```suggestion
	defer func() {
		bc.db.FinishStateMigration(returnErr == nil)
		bc.wg.Done()
	}()
```

File: storage/database/db_manager.go [hunk 2]
```diff
@@ -476,12 +476,76 @@ func (dbm *databaseManager) NewBatch(dbEntryType DBEntryType) Batch {
 		defer dbm.lockInMigration.RUnlock()
 
 		if dbm.inMigration {
-			return dbm.GetStateTrieMigrationDB().NewBatch()
+			newDBBatch := dbm.getDatabase(StateTrieMigrationDB).NewBatch()
+			oldDBBatch := dbm.getDatabase(StateTrieDB).NewBatch()
+			return NewStateTrieDBBatch([]Batch{oldDBBatch, newDBBatch})
```
Comment (aidan-kwon) [inline storage/database/db_manager.go:481]: Is there any place we need to write data to both of oldDB and migrationDB?

Comment (KimKyungup) [reply storage/database/db_manager.go:481]: For recovering old database, this PR introduce this batch in migration.
This is used for `trieDB.Commit` / `trieDB.Cap`.

File: storage/database/db_manager.go [hunk 3]
```diff
@@ -476,12 +476,76 @@ func (dbm *databaseManager) NewBatch(dbEntryType DBEntryType) Batch {
...
+			maxSize = batch.ValueSize()
+		}
+	}
+
+	return maxSize
+}
+
+// Write passes the list of batch tasks to taskCh so batch can be processed
+// by underlying workers. Write waits until all workers return the result.
+func (stdBatch *stateTrieDBBatch) Write() error {
```
Comment (winnie-byun) [inline storage/database/db_manager.go:524]: There are functions such as `WriteBatches` and `WriteBatchesOverThreshold` and this seems it is duplicated code. What do you think about modifying the functions and calling the functions here?

File: storage/database/db_manager.go [hunk 4]
```diff
@@ -476,12 +476,76 @@ func (dbm *databaseManager) NewBatch(dbEntryType DBEntryType) Batch {
...
+		}
+	}
+
+	return errResult
+}
+
+// ValueSize is called to determine whether to write batches when it exceeds
+// certain limit. partitionedDB returns the largest size of its batches to
+// write all batches at once when one of batch exceeds the limit.
+func (stdBatch *stateTrieDBBatch) ValueSize() int {
```
Comment (aidan-kwon) [inline storage/database/db_manager.go:511]: How about returning `-1` if values sizes of batches have different values.

File: storage/database/db_manager.go (hunk 3 above)
Comment (KimKyungup) [reply storage/database/db_manager.go:524]: I think this stateTrieDBBatch is the lower level than `WriteBatches` and `WriteBatchesOverThreshold`.
They call state trie batch write and use `stateTrieDBBatch` in migration.
So, They are not duplicated.

Comment (winnie-byun) [reply storage/database/db_manager.go:524]: I don't get that you said it is **other level**. stdBatch can call `WriteBatches` like this without an inifinite loop : 
```
func (stdBatch *stateTrieDBBatch) Write() error {
	_, err := WriteBatches(stdBatch.batches...)
	return err
}
```
If you don't want to change the definition of `WriteBatches`, a new function can be defined :
```
func WriteBatchesParallel(batches ...Batch) (int, error) {
   ...
}

func (stdBatch *stateTrieDBBatch) Write() error {
	_, err := WriteBatchesParallel(stdBatch.batches...)
	return err
}
```
I still think the code can be reused in some place else and this code limits its usage in `stdBatch`.
What do you think? :)

File: blockchain/state_migration.go [hunk 5]
```diff
@@ -84,9 +84,12 @@ func (bc *BlockChain) concurrentRead(db *statedb.Database, quitCh chan struct{},
 	}
 }
 
-func (bc *BlockChain) migrateState(rootHash common.Hash) error {
+func (bc *BlockChain) migrateState(rootHash common.Hash) (returnErr error) {
```
Comment (aidan-kwon) [inline blockchain/state_migration.go:87]: The return of `migrateState` is not used in anywhere.

**Comment by aidan-kwon** [comment]
Doesn't `bc.prepareStateMigration` need to be `true` when migration is stopped?

File: blockchain/state_migration.go (hunk 5 above)
Comment (KimKyungup) [reply blockchain/state_migration.go:87]: Right. But, in the defer function, it can be used well.
Is it ok?

File: storage/database/db_manager.go (hunk 3 above)
Comment (KimKyungup) [reply storage/database/db_manager.go:524]: I reflected. Thanks. :) @winnie-byun PTAL.

**Comment by KimKyungup** [comment]
> Doesn't `bc.prepareStateMigration` need to be `true` when migration is stopped?

@aidan-kwon 
If we set it `true` again, it will retry again.
I think setting again it and retry after stop and recover by some error is weired.

In that case, It think user should know the status and consider to report or retry his/herself.

File: storage/database/db_manager.go (hunk 3 above)
Comment (winnie-byun) [reply storage/database/db_manager.go:524]: Thanks for considering. LGTM :)

**Comment by KimKyungup** [comment]
PTAL. :)

**Comment by KimKyungup** [comment]
@aidan-kwon PTAL. :)

**Comment by KimKyungup** [comment]
I rebased. PTAL. :)

File: storage/database/db_manager.go [hunk 6]
```diff
@@ -476,12 +476,62 @@ func (dbm *databaseManager) NewBatch(dbEntryType DBEntryType) Batch {
...
+		if batch.ValueSize() > maxSize {
+			maxSize = batch.ValueSize()
+		}
+	}
+
+	return maxSize
+}
+
+// Write passes the list of batch tasks to taskCh so batch can be processed
+// by underlying workers. Write waits until all workers return the result.
```
Comment (ehnuje) [inline storage/database/db_manager.go:523]: The above comment seems unrelated to this function. Maybe copied from partitioned database batch?

File: storage/database/db_manager.go [hunk 7]
```diff
@@ -476,12 +476,62 @@ func (dbm *databaseManager) NewBatch(dbEntryType DBEntryType) Batch {
...
+		if err := batch.Put(key, value); err != nil {
+			errResult = err
+		}
+	}
+
+	return errResult
+}
+
+// ValueSize is called to determine whether to write batches when it exceeds
+// certain limit. partitionedDB returns the largest size of its batches to
```
Comment (ehnuje) [inline storage/database/db_manager.go:509]:
```suggestion
// certain limit. stdBatch returns the largest size of its batches to
```

File: storage/database/db_manager.go (hunk 6 above)
Comment (KimKyungup) [reply storage/database/db_manager.go:523]: Right. I modified. :) Thanks.

File: storage/database/interface.go [hunk 8]
```diff
@@ -93,6 +93,38 @@ func WriteBatches(batches ...Batch) (int, error) {
...
+		}(batch)
+	}
+
+	var bytes int
+	var errResult error
+
+	for range batches {
+		rst := <-resultCh
+		if rst.err != nil {
+			errResult = rst.err
```
Comment (ehnuje) [inline storage/database/interface.go:120]: Don't we need to leave a log for each error? Since only the last error will be returned and logged.

Comment (KimKyungup) [reply storage/database/interface.go:120]: I think it's ok to know one error even if there are several errors.
But if we know all error. How about this code?
But I think I don't need this code. What do you think? :)
```suggestion
			errResult = fmt.Errorf("%v - %v", errResult, rst.err)
```

File: blockchain/state_migration.go [hunk 9]
```diff
@@ -209,8 +209,7 @@ func (bc *BlockChain) migrateState(rootHash common.Hash) error {
 	checkElapsed := time.Since(startCheck)
 	logger.Info("State migration : Consistency check is done", "elapsed", checkElapsed)
 
-	bc.db.FinishStateMigration()
-	logger.Info("State migration : Finished")
+	logger.Info("completed state migration")
```
Comment (ehnuje) [inline blockchain/state_migration.go:212]: How about merging two lines of logs instead of individual lines?

File: storage/database/interface.go (hunk 8 above)
Comment (ehnuje) [reply storage/database/interface.go:120]: If so, we can immediately return the error once we receive it. As batch write will be done anyway as they are executed by go-routine. Also the channel will be garbage-collected. We don't need to wait to check the whole result.
```suggestion
			return rst.err
```

Comment (KimKyungup) [reply storage/database/interface.go]: Good point. I reflected. :)

File: blockchain/state_migration.go (hunk 9 above)
Comment (KimKyungup) [reply blockchain/state_migration.go]: I reflected. :)

**Comment by KimKyungup** [comment]
**Test Result**
I think we need to improve the speed of checking state trie.

- Elapsed Time
  - Copy : 10h10m59.305838312s
  - Check : 34h4m56.846618626s

```
INFO[06/17,23:44:33 +09] [5] State migration : Copy is done            totalRead=363656866 totalCommitted=363656866 totalElapsed=10h10m59.305838312s committed per second=9919.906
INFO[06/19,09:49:30 +09] [5] State migration : Consistency check is done  elapsed=34h4m56.846618626s
INFO[06/19,09:49:30 +09] [5] completed state migration
```


- Size
```
57G	./body
2.8G	./misc
232K	./bridgeservice
62G	./header
6.7G	./txlookup
42G	./receipts
70G	./statetrie_migrated_29307517
239G	.
```
